Improve linking

3c1u opened the pull request:
It is quite a hardship to link Live2DCubismCore properly because the linking options that Live2D offers are completely broken:

* macOS
  - ~The static library `libLive2DCubismCore.a` is a fat binary; cannot be linked directly from the linker.~ 64-bit only on macOS
  - The "dynamic library" `Live2DCubismCore.bundle` is not in the normal .dylib format.
     This means that, it cannot be linked as a dynamic library, and it requires to use deprecated APIs (since 10.5) to load (`NSCreateObjectFileImageFromFile`, etc...).

* Linux
  - ~Since the static library is not compiled with `-fPIC` option, the static linking will fail.~ Fixed on v4.

To solve this problem, I made some workarounds:
* Switchable linking options: `"static"`, `"dylib"`
* ~Convert the fat library to thinned libraries on macOS target.~
* Panics if the target is macOS and trying to link dynamically.
* ~Warn if the target is Linux and trying to link statically.~

Veykril commented:
Nice, this was long overdue I never got to test any other platform than windows since I didnt have access to any other for the time being. There are two things I'd like to discuss here. One being that I don't like having two feature flags for this. It should be enough to have a single one like `static-link` meaning if its enabled we link statically, if its not we link dynamically. The second being regarding the thinning of the macos fat library, if i got this correctly the buildscript overwrites the library with thin one now which isn't the best idea i think. It would probably be nicer in having it output the thin lib in the `target` directory and link it there, this should be doable through the environment flags cargo sets but I am not certain since I haven't done this before.

Note: I accidentally marked it as ready for review. 😅

3c1u commented:
I modified `macos_slim_lib` and made the `third-party` directory intact in the function. I changed the feature flag.

Also reported this problem to Live2D: [macOSおよびLinux環境でのリンクについて](https://forum.live2d.com/discussion/1621/macos%E3%81%8A%E3%82%88%E3%81%B3linux%E7%92%B0%E5%A2%83%E3%81%A7%E3%81%AE%E3%83%AA%E3%83%B3%E3%82%AF%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6)

3c1u commented:
By the way, the dynamic library for macOS can be produced by the following commands:

```
$ gcc -dynamiclib -all_load -o libLive2DCubismCore.dylib \
   libLive2DCubismCore.a -arch x86_64
$ mv libLive2DCubismCore.dylib libLive2DCubismCore64.dylib
$ gcc -dynamiclib -all_load -o libLive2DCubismCore.dylib \
   libLive2DCubismCore.a -arch i686
$ mv libLive2DCubismCore.dylib libLive2DCubismCore32.dylib
$ lipo -create -arch x86_64 libLive2DCubismCore64.dylib \
   -arch i686 libLive2DCubismCore32.dylib \
   -output libLive2DCubismCore.dylib
```

I don't want to include this to a build script, though.

Veykril commented:
I don't feel to good about adding that macOS command chain you showed there either, I hope they'll fix that.
Otherwise this seems good to me now, tell me if you are done with it and I will merge it.

3c1u commented:
Static linking on macOS/Linux is fixed on v4! They also promised to include macOS .dylib format.

Veykril commented:
That's great to hear, I assume this PR is ready to merge then?

3c1u commented:
Yep, it's ready.

Veykril commented on cubism-core-sys/build.rs:
```diff
@@ -29,6 +31,17 @@ fn main() {
             lib_dir.push("140");
         },
         ("apple", "darwin") => {
+            if cfg!(feature = "dylib") {
```
This should be 
```rust
if !cfg!(feature = "static-link") {
```
or
```rust
if linking_strategy != "static" {
```

3c1u replied:
Oh, It won't work because there is no "dylib" feature now.

I'd prefer
```rust
if linking_strategy == "dylib" {
```
but `linking_strategy != "static"` has smaller comparing cost so I'll choose this.

Veykril commented on cubism-core-sys/build.rs:
```diff
@@ -1,23 +1,25 @@
+        lib_dir
+    } else {
+        // TODO: this won't appear unless the compilation fails.
+        println!(
```
I think we can just panic here as well, the compilation will fail at the linking stage anyways so panicking here probably causes cleaner output?

3c1u replied:
Some people will prefer linking libraries manually (e.g. placing libraries to the default library directory, writing build.rs by their own) to specifying libraries by setting environment variables.

Veykril replied:
ah yep makes sense

Veykril commented:
found one small thing and another question
